lib/v1.0/C/secure_pprz_transport.c

```diff
@@ -0,0 +1,316 @@
+ * See https://wiki.paparazziuav.org/wiki/Messages_Format
+ * for more details
+ *
+ * Pprz frame:
+ *
+ * |STX|length|... payload=(length-4) bytes ...|Checksum A|Checksum B|
+ *
+ * where checksum is computed over length and payload:
+ * @code
+ * ck_A = ck_B = length
+ * for each byte b in payload
+ *     ck_A += b;
+ *     ck_b += ck_A;
+ * @endcode
+ */
+
+#include <inttypes.h>
+#include "pprzlink/secure_pprz_transport.h"
+
+// crypto
+#include "pprzlink/hacl-c/Chacha20Poly1305.h"
+//#include "generated/keys.h"
+// will be in keys.h
+// for DEBUG Crypto, both keys are the same
+#define UAV_RX_KEY { 0x70, 0x3, 0xAA, 0xA, 0x8E, 0xE9, 0xA8, 0xFF, 0xD5, 0x46, 0x1E, 0xEC, 0x7C, 0xC1, 0xC1, 0xA1, 0x6A, 0x43, 0xC9, 0xD4, 0xB3, 0x2B, 0x94, 0x7E, 0x76, 0xF9, 0xD8, 0xE8, 0x1A, 0x31, 0x5D, 0xA8 }
+#define UAV_TX_KEY { 0x70, 0x3, 0xAA, 0xA, 0x8E, 0xE9, 0xA8, 0xFF, 0xD5, 0x46, 0x1E, 0xEC, 0x7C, 0xC1, 0xC1, 0xA1, 0x6A, 0x43, 0xC9, 0xD4, 0xB3, 0x2B, 0x94, 0x7E, 0x76, 0xF9, 0xD8, 0xE8, 0x1A, 0x31, 0x5D, 0xA8 }
+
+// PPRZ parsing state machine
+#define UNINIT      0
+#define GOT_STX     1
+#define GOT_LENGTH  2
+#define GOT_PAYLOAD 3
+#define GOT_CRC1    4
+
+#define msg_put_byte(_t,_b) {\
+    _t.msg[_t.msg_idx] = _b;\
+    _t.msg_idx++;\
+}
+
+#define buffer_put_byte(_t,_b) {\
+    _t->tx_buffer[_t->tx_idx] = _b;\
+    _t->tx_idx++;\
+}
+
+static void accumulate_checksum(struct spprz_transport *trans, const uint8_t byte)
+{
+  trans->ck_a_tx += byte;
+  trans->ck_b_tx += trans->ck_a_tx;
+}
+
+static void put_bytes(struct spprz_transport *trans, struct link_device *dev __attribute__((unused)), long fd __attribute__((unused)),
+                      enum TransportDataType type __attribute__((unused)), enum TransportDataFormat format __attribute__((unused)),
+                      const void *bytes, uint16_t len)
+{
+  const uint8_t *b = (const uint8_t *) bytes;
+  int i;
+  for (i = 0; i < len; i++) {
+    // note we are putting bytes indiscriminately into the msg buffer
+    msg_put_byte(trans->tx_msg, b[i]);
+  }
+}
+
+static void put_named_byte(struct spprz_transport *trans, struct link_device *dev __attribute__((unused)), long fd __attribute__((unused))  ,
+                           enum TransportDataType type __attribute__((unused)), enum TransportDataFormat format __attribute__((unused)),
+                           uint8_t byte, const char *name __attribute__((unused)))
+{
+  // note we are putting bytes indiscriminately into the msg buffer
+  msg_put_byte(trans->tx_msg, byte);
+}
+
+static uint8_t size_of(struct spprz_transport *trans __attribute__((unused)), uint8_t len)
+{
+  // message length: payload + protocol overhead (STX + len + ck_a + ck_b = 4) + crypto overhead (4 byte counter, 16 byte tag)
+  return len + PPRZ_HEADER_LEN + PPRZ_CRYPTO_OVERHEAD;
+}
+
+static void start_message(struct spprz_transport *trans, struct link_device *dev __attribute__((unused)), long fd __attribute__((unused)), uint8_t payload_len)
```

Comment:
we cannot send a message until it is all processed and encrypted, hence we use an intermediate buffer to store data and quietly assume that only one process can write the data